### PR TITLE
Support text filenames and no-PATHEXT no-extension

### PR DESCRIPTION
**Pull request by asottile**

### CODE REVIEW

**Comment by asottile** [comment]
Sorry I did two things in one commit.  I was writing a test for the first thing and hit the second thing :)

Some notes about this commit:
- Your runtime code doesn't depend on setuptools so I removed it from install_requires -- you could put it in setup_requires, but you're importing it at the top of the file so that's too late already :)
- I added a dependency on six for `six.string_types`

File: test/test_parse_shebang.py [hunk 1]
```diff
@@ -146,3 +148,27 @@ def test_throw_runtime_error_when_weird_object_passed(self):
...
             parse(object())
+
+    # suppress(no-self-use)
+    def test_unicode(self):
+        """A filename can be text which in python 2 is not `str`."""
+        tmpdir = mkdtemp()
+        self.addCleanup(shutil.rmtree, tmpdir)
+
+        filename = os.path.join(tmpdir, u"foo.sh")
+        with io.open(filename, "w") as f:
```
Comment (smspillaz) [inline test/test_parse_shebang.py]: This is really just a matter of curiosity - what is the difference between using the `io` module here and just using `open`? I think this is the first time I've actually come across the module (forgive me, the python standard library is huge).

Having a quick skim over the documentation, it looks like the module brings whatever IO semantics python 3 had to python 2 in an opt-in manner. Is that right?

File: parseshebang/__init__.py [hunk 2]
```diff
@@ -45,8 +47,11 @@ def parse(file_to_parse):  # suppress(unused-function)
     """
     if hasattr(file_to_parse, "read"):
         return _parse(file_to_parse)
-    elif isinstance(file_to_parse, str):
-        path_ext = os.environ.get("PATHEXT", "").split(os.pathsep)
+    elif isinstance(file_to_parse, six.string_types):
+        if "PATHEXT" in os.environ:
+            path_ext = os.environ["PATHEXT"].split(os.pathsep)
```
Comment (smspillaz) [inline parseshebang/__init__.py]: This looks fine. The only suggestion I have is that if we're going to use `frozenset` it should probably be used consistently (e.g., convert the return value of `os.environ["PATHEXT"].split(os.pathsep)` to be a `frozenset` too.

**Comment by smspillaz** [comment]
> Your runtime code doesn't depend on setuptools so I removed it from install_requires -- you could put it in setup_requires, but you're importing it at the top of the file so that's too late already :)

Oh, I had a question to ask about this.

So bit that you caught (setuptools being in `install_requires`) was definitely wrong. Thanks for catching that.

However, I have seen it in `setup_requires` for a few projects. I'm not sure if this is a convention of some sort that needs to be followed. I know that setuptools gets updated out of cycle with python and so sometimes your setup script might depend on a newer version. However, I've always found this to be perplexing in the sense that `setuptools` is imported at the top of the file so surely its too late at that point? Unless setuptools has some magic to import setuptools, ignore failing imports, install the stuff in `setup_requires` and then reload everything.

File: parseshebang/__init__.py (hunk 2 above)
Comment (asottile) [inline parseshebang/__init__.py]: Switched to empty tuple instead -- it has the same runtime characteristics anyway and maintains the same type as splitting the environment variable :)

File: test/test_parse_shebang.py (hunk 1 above)
Comment (asottile) [inline test/test_parse_shebang.py]: Indeed that is correct.  Whenever I write 2+3 code I use the `io` module so I'm always dealing with the same types on both sides of the fence

**Comment by asottile** [comment]
Setuptools being in `setup_requires` is also incorrect but the background of it is somewhat interesting.

`distutils` and `setuptools` way back when both existed and served the same purpose (packaging things!) in the era of `easy_install` (no `pip` to be found).

Since `distutils` was stuck in the stdlib, `setuptools` built tons of improvements and was installable and upgradable.  This often lead to code in setup.py looking like:

```python
try:
    from setuptools import setup
except ImportError:
    from distutils import setup

setup(
    # ...
    setup_requires=['setuptools'],
)
```

When `setuptools` was not there, `distutils` would install it as part of `setup_requires` and then it would get used for the rest of the setup steps.

As pip became more and more ubiquitous, setuptools did as well.  In the modern day, it's pretty safe to assume that a python installation has access to setuptools and so the `try: except:` was abandoned.   The `setup_requires` bit stuck around mostly by mistake (and because it sounds eerily similar to `install_requires`).  Lots of copy paste + cargo cult and you end up with it everywhere!

**Comment by smspillaz** [comment]
Thanks for answering those questions! That's actually really helpful to know - I'll make sure to update the rest of my modules on that advice.

**Pull request merged**